### PR TITLE
MIDI-35: Forward diffusion process

### PR DESCRIPTION
**Pull request by JasiekKaczmarczyk**
[Build a 1D forward diffusion process](https://sacrebleu.atlassian.net/browse/MIDI-35)

Implementation of forward diffusion process for ECG signals.

Here are some examples of forward diffusion:
forward_diffusion_cosine

forward_diffusion_sigmoid

### CODE REVIEW

**Comment by roszcz** [comment]
Looks like it's noise already for step 0 - is this because you started with fragments without any beats? If so, can you also show the results for samples with clean ecg? :)

File: ecg_segmentation_dataset.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        super().__init__()
+
+        features = Features(
+            {
+                "record_id": Value(dtype="string"),
+                "signal": Array2D(dtype="float32", shape=(2, 1000)),
+                "mask": Array2D(dtype="int8", shape=(1, 1000)),
+            }
+        )
+        self.data = load_dataset(huggingface_path, features=features, split=split)
```
Comment (roszcz) [inline ecg_segmentation_dataset.py:17]:
```suggestion
        self.data = load_dataset(huggingface_path, split=split)
```

I think it's not necessary to provide features here if we created the dataset including information about datatype? 🤔 Can you check please?

File: screenshots/forward_diffusion_cosine.png
Comment (roszcz) [inline screenshots/forward_diffusion_cosine.png:1]: I think it's better not to commit screenshots to the repository, but use the github storage for images - just drag and drop them into PR description or comment

File: tests/forward_diffusion_test.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    train_dataloader = DataLoader(train_ds, batch_size=batch_size, num_workers=num_workers)
+    val_dataloader = DataLoader(val_ds, batch_size=batch_size, num_workers=num_workers)
+    test_dataloader = DataLoader(test_ds, batch_size=batch_size, num_workers=num_workers)
+
+    return train_dataloader, val_dataloader, test_dataloader
+
+
+if __name__ == "__main__":
+    # load data
+    _, _, test_dataloader = preprocess_dataset("roszcz/ecg-segmentation-ltafdb", 4, 1)
```
Comment (roszcz) [inline tests/forward_diffusion_test.py]:
```suggestion
    _, _, test_dataloader = preprocess_dataset(
      dataset_name="roszcz/ecg-segmentation-ltafdb",
      batch_size=4,
      num_workers=1,
    )
```

It's much easier to understand with named arguments

**Comment by JasiekKaczmarczyk** [comment]
Here are examples of signals with beats:
forward_diffusion_cosine_heartbeats

forward_diffusion_sigmoid_heartbeats

I think it would be better to use sigmoid schedule because it destroys data slower.

File: tests/forward_diffusion_test.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    records = next(iter(test_dataloader))
+    signal = records["signal"]
+
+    schedule_type = "sigmoid"
+
+    # initialize forward diffusion
+    fdiff = ForwardDiffusion(beta_start=0.0001, beta_end=0.02, timesteps=256, schedule_type=schedule_type)
+
+    # timesteps for visualization
+    t = torch.tensor([0, 63, 127, 191, 255], dtype=torch.long)
```
Comment (roszcz) [inline tests/forward_diffusion_test.py]:
```suggestion
    t = torch.tensor([0, 10, 30, 63, 127, 255], dtype=torch.long)
```

63 steps seems like a lot, can you post some images for the earlier steps as well?

**Comment by roszcz** [comment]
> I think it would be better to use sigmoid schedule because it destroys data slower.

That's assuming the slower destruction is going to lead to better results - this would also be my guess, but I think we'll have to find out empirically 🔎

File: ecg_segmentation_dataset.py (hunk 1 above)
Comment (JasiekKaczmarczyk) [reply ecg_segmentation_dataset.py:17]: I think it's not necessary, but I thought that it would make code a little bit easier to understand if dtypes are specified.

Comment (roszcz) [reply ecg_segmentation_dataset.py:17]: If it makes it easier to read for you than it's ok to leave it in. If you ever need to check the datatypes of a dataset you can also use `dataset.features` attribute.

File: tests/forward_diffusion_test.py (hunk 3 above)
Comment (JasiekKaczmarczyk) [reply tests/forward_diffusion_test.py]: forward_diffusion_cosine_heartbeats
forward_diffusion_sigmoid_heartbeats

**Review by roszcz — APPROVED** [review]
